### PR TITLE
[ThemeBundle] Refactored low-level theme management, introduced Composer configuration sharing support

### PR DESCRIPTION
**Pull request by pamil**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes (but within development version 0.17)
| Deprecations? | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Based on #3945. This PR adds support for sharing `composer.json` being like `theme.json` - a central, configuration file for theme. Also, low-level theme-management was refactored, allowing us to be more flexible and more strict at the same time - e.g. you can provide theme configuration from every source you want to, but parsing configuration is much more strict that is was before. To be continued... feel free to review!

TODO:
  - [x] Authors support
  - [ ] Documentation
  - [ ] Elastic configuration for different configuration sources

Naming changes:
 - name -> title
 - slug -> name (following Composer convention)
 - parentsSlugs -> parentsNames

### CODE REVIEW

File: src/Sylius/Bundle/ThemeBundle/Loader/ConfigurationProcessingLoader.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<?php
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/ThemeBundle/Loader/ConfigurationProcessingLoader.php]: Missing license block :(

File: src/Sylius/Bundle/ThemeBundle/Loader/ConfigurationProcessorInterface.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+interface ConfigurationProcessorInterface
+{
+    /**
+     * @param array $configs An array of configuration arrays
+     *
+     * @return array The processed configuration array
```
Comment (Zales0123) [inline src/Sylius/Bundle/ThemeBundle/Loader/ConfigurationProcessorInterface.php]: Isn't the comment redundant?

File: src/Sylius/Bundle/ThemeBundle/Loader/ConfigurationProcessingLoader.php (hunk 1 above)
Comment (pamil) [inline src/Sylius/Bundle/ThemeBundle/Loader/ConfigurationProcessingLoader.php]: Added :)

**Comment by pjedrzejewski** [comment]
@pamil I'd say we leave "Elastic configuration for different configuration sources" for later. Let's focus on adding the docs and releasing the bundle. Extra features will come. :)

**Comment by pamil** [comment]
@pjedrzejewski yeah, that's something that we can introduce in further stage of development, nevertheless I modified default `sylius.locations` configuration option to `sylius.sources.filesystem.locations` so we can add another sources in the future without modifying the configuration structure.

**Comment by pjedrzejewski** [comment]
Thanks Kamil! :+1:

**Pull request merged**